python/ray/data/_internal/datasource/kafka_datasource.py

```diff
@@ -0,0 +1,568 @@
+                            )
+
+                            # Poll for a batch of messages from Kafka
+                            msg_batch = consumer.poll(
+                                timeout_ms=min(remaining_timeout_ms, 10000),
+                            )
+
+                            if not msg_batch:
+                                # No more messages available right now
+                                # Yield any accumulated records and exit the loop
+                                continue
+
+                            messages = msg_batch.get(topic_partition, [])
+
+                            for msg in messages:
+                                # Check if we've reached the end offset (for bounded reads)
+                                # Use >= for exclusive end_offset (don't include end_offset message)
+                                if end_off is not None and msg.offset >= end_off:
+                                    partition_done = True
+                                    break
+
+                                # Extract all message metadata into a flat record
+                                headers_dict = _convert_headers_to_dict(msg.headers)
+
+                                records.append(
+                                    {
+                                        "offset": msg.offset,
+                                        "key": msg.key,
+                                        "value": msg.value,
+                                        "topic": msg.topic,
+                                        "partition": msg.partition,
+                                        "timestamp": msg.timestamp,
+                                        "timestamp_type": msg.timestamp_type,
+                                        "headers": headers_dict,
+                                    }
+                                )
+                                records_read += 1
+
+                                # Yield incrementally when we hit batch size
+                                if len(records) >= KafkaDatasource.BATCH_SIZE_FOR_YIELD:
+                                    table = pa.Table.from_pylist(records)
+                                    output_buffer.add_block(table)
+                                    if output_buffer.has_next():
+                                        yield output_buffer.next()
+                                    records = []  # Clear for next batch
+
+                        # Yield any remaining records
+                        if records:
+                            table = pa.Table.from_pylist(records)
+                            output_buffer.add_block(table)
+                            if output_buffer.has_next():
+                                yield output_buffer.next()
+
+                        output_buffer.finalize()
+                        if output_buffer.has_next():
+                            yield output_buffer.next()
```

Comment:
I think this is eequivalent?
```suggestion
                        # Yield any remaining records
                        if records:
                            table = pa.Table.from_pylist(records)
                            output_buffer.add_block(table)

                        output_buffer.finalize()
                        while output_buffer.has_next():
                            yield output_buffer.next()
```